build.py

```diff
@@ -0,0 +1,368 @@
+- clean             removes build artifacts from the project directories
+- build-sql         constructs the sql files for the extension
+- clean-sql         removes sql file artifacts from the project directories
+- install-sql       installs the sql files into the postgres installation
+- uninstall-sql     removes the sql extension from the postgres installation
+- install-prior-py  installs the python package and dependencies from previously released versions
+- install-py        installs the python package and dependencies for the current version
+- clean-py          removes python build artifacts from the project directories
+- uninstall-py      removes the python package and dependencies from the system
+- test              runs the unit tests in the docker database
+- docker-build      builds the development docker image
+- docker-run        launches a container in docker using the docker image
+- docker-stop       stops the container
+- docker-rm         deletes the development container"""
+
+
+def versions() -> list[str]:
+    return ["0.4.0", "0.3.0", "0.2.0", "0.1.0"]  # ADD NEW VERSIONS TO THE FRONT OF THIS LIST! STAY SORTED PLEASE
+
+
+def this_version() -> str:
+    return versions()[0]
+
+
+def prior_versions() -> list[str]:
+    return versions()[1:] if len(versions()) > 1 else []
+
+
+def project_dir() -> Path:
+    return Path(__file__).resolve().absolute().parent
+
+
+def sql_dir() -> Path:
+    return project_dir().joinpath("sql")
+
+
+def src_dir() -> Path:
+    return project_dir().joinpath("src")
+
+
+def incremental_sql_dir() -> Path:
+    return sql_dir().joinpath("incremental")
+
+
+def idempotent_sql_dir() -> Path:
+    return sql_dir().joinpath("idempotent")
+
+
+def idempotent_sql_files() -> list[Path]:
+    paths = [x for x in idempotent_sql_dir().glob("*.sql")]
+    paths.sort()
+    return paths
+
+
+def incremental_sql_files() -> list[Path]:
+    paths = [x for x in incremental_sql_dir().glob("*.sql")]
+    paths.sort()
+    return paths
+
+
+def output_sql_file() -> Path:
+    return sql_dir().joinpath(f"ai--{this_version()}.sql")
+
+
+def build_sql_control_file() -> None:
+    ctl_file = sql_dir().joinpath("ai.control")
+    content = ctl_file.read_text()
+    lines = []
+    for line in content.splitlines(keepends=True):
+        if line.startswith("default_version"):
+            lines.append(f"default_version='{this_version()}'\n")
+        else:
+            lines.append(line)
+    ctl_file.write_text("".join(lines))
+
+
+def build_incremental_sql_file(input_file: Path) -> str:
+    template = sql_dir().joinpath("migration.sql").read_text()
+    migration_name = input_file.name
+    migration_body = input_file.read_text()
+    version = this_version()
+    return template.format(migration_name=migration_name, migration_body=migration_body, version=version)
+
+
+def build_idempotent_sql_file(input_file: Path) -> str:
+    inject = f"""
+    if "ai.python_lib_dir" not in GD:
```

Comment:
how does this work when an extension is updated. Do we clear this in the update script? Is it possible to remove the sitedir?

Reply:
When we release version 0.4.0, we will check in all the `ai--*0.4.0.sql` scripts which will have the path to the 0.4.0 python directory hardcoded in them (courtesy of this function).
Then, we put 0.5.0 at the front of the versions list, and the next time we build we get `ai--*0.5.0.sql` scripts with the path hardcoded to the new 0.5.0 python directory.
When you do an `alter extension update` all the 0.4.0 functions will get replaced with 0.5.0 functions with the new path. You'll need to start a new database session after the upgrade.

Reply:
maybe we could improve it to remove the need for starting a new session though

Reply:
i don't see a "remove site dir" function at quick glance

Reply:
https://github.com/timescale/pgai/pull/47/commits/c0eef3aa452e054ed0d84160daf938fe0954419d

Will now bail if the version has changed.